url49584c1d41_Arrem_GGS_tgz-pJ8-com_gamezgalaxy_test_console_MainJ8/src/com/gamezgalaxy/GGS/server/Server.java

```diff
@@ -0,0 +1,521 @@
+		ch.addCommand(new Unban());
+
+	}
+	
+	private static String LetterOrNumber(String string) {
+		final String works = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz1234567890";
+		String finals = "";
+		boolean change = true;
+		for (char c : string.toCharArray()) {
+			for (char x : works.toCharArray()) {
+				if (x == c) {
+					change = false;
+					break;
+				}
+			}
+			if (change)
+				finals += works.toCharArray()[new Random().nextInt(works.toCharArray().length)];
+			else
+				finals += c;
+			change = true;
+		}
+		return finals;
+	}
+
+	/**
+	 * Stop the server, this will kick all the players on the server
+	 * and stop all server services.
+	 * @throws InterruptedException
+	 *                             if any thread has interrupted the current thread. The interrupted status of the current thread is cleared when this exception is thrown.
+	 * @throws IOException
+	 *                    If there is a problem saving the levels that are loaded
+	 */
+	public void Stop() throws InterruptedException, IOException {
+		if (!Running)
+			return;
+		Running = false;
+		Log("Stopping server...");
+		for(Player p : players)
+		{
+			p.sendMessage("Stopping server...");
+		}
+		for (Level l : this.getLevelHandler().getLevelList()) {
+			if (l != MainLevel)
+				l.Unload(this);
+		}
+		MainLevel.Save();
+		tick.join();
+		logger.Stop();
+		heartbeater.stop();
+		for(Player p : players) // Implementing this again because I want to stop everything first.
+		{						// The idea is that at the begining of the "stopping" session, the player receives that the server is stopping.
+								// Then after everything has stopped but the server itself, the player is kicked from the server before actually shutdown.
+			p.Kick("Server shut down. Thanks for playing!"); // Kicking so we can place a message that the server was stopped.
+
+		}
+		pm.StopReading();
+	}
+	
+	/**
+	 * Log something to the logs
+	 * @param log
+	 */
+	public void Log(String log) {
+		logger.Log(log);
+	}
+	
+	/**
+	 * Get the logger object
+	 * @return
+	 *        The {@link Logger} object
+	 */
+	public Logger getLogger() {
+		return logger;
+	}
+
+	/**
+	 * Add a task to be called every 10 milliseconds
+	 * @param t
+	 *         The {@link Tick} object to call
+	 */
+	public  void Add(Tick t) {
+		synchronized(ticks) {
+			if (!ticks.contains(t))
+				ticks.add(t);
+		}
+	}
+	
+	/**
+	 * Remove a task from the Tick list
+	 * @param t
+	 *         The {@link Tick} object to remove
+	 */
+	public void Remove(Tick t) {
+		synchronized(ticks) {
+			if (ticks.contains(t))
+				ticks.remove(t);
+		}
+	}
+
+	private class Ticker extends Thread {
+
+		@Override
+		public void run() {
+			while (Running) {
+				for (int i = 0; i < ticks.size(); i++) {
+					Tick t = ticks.get(i);
+					try {
+						t.tick();
+					} catch (Exception e) {
+						Log("ERROR TICKING!");
+						e.printStackTrace();
+					}
+				}
+				try {
+					Thread.sleep(10);
+				} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)